lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableObserverContainer.cs

```
@@ -0,0 +1,127 @@
+        {
+            _typeMap[typeof(T)] = observer;
+            return Disposable.Create(() => _typeMap.TryRemove(typeof(T), out observer));
+        }
+
+        /// <summary>
+        /// Look up the IObserver for the registered IPEndPoint or event type 
+        /// and execute the IObserver.
+        /// </summary>
+        /// <param name="transportEvent">The incoming remote event</param>
+        public void OnNext(TransportEvent<IWritableRemoteEvent<T>> transportEvent)
+        {
+            IWritableRemoteEvent<T> remoteEvent = transportEvent.Data;
+            remoteEvent.LocalEndPoint = transportEvent.Link.LocalEndpoint;
+            remoteEvent.RemoteEndPoint = transportEvent.Link.RemoteEndpoint;
+            T value = remoteEvent.Value;
+            bool handled = false;
+
+            IObserver<T> observer1;
+            IObserver<IRemoteMessage<T>> observer2;
+            if (_universalObserver != null)
+            {
+                _universalObserver.OnNext(value);
+                handled = true;
+            }
+            if (_endpointMap.TryGetValue(remoteEvent.RemoteEndPoint, out observer1))
+            {
+                // IObserver was registered by IPEndpoint
+                observer1.OnNext(value);
+                handled = true;
+            } 
+            else if (_typeMap.TryGetValue(value.GetType(), out observer2))
+            {
+                // IObserver was registered by event type
+                IRemoteIdentifier id = new SocketRemoteIdentifier(remoteEvent.RemoteEndPoint);
+                IRemoteMessage<T> remoteMessage = new DefaultRemoteMessage<T>(id, value);
+                observer2.OnNext(remoteMessage);
+                handled = true;
+            }
+
+            if (!handled)
+            {
+                throw new WakeRuntimeException("Unrecognized Wake RemoteEvent message");
+            }
+        }
+
+        public void OnError(Exception error)
```

Please throw the `error` here. Otherwise we might eat exceptions.